### ISSUE TITLE
Problem with variable types.

### ISSUE THREAD

**Issue by hecperleo** [stonier/ecl_core#60]
Hello! I'm starting to work with ROS and there are bugs that I can't fix. 

I'm trying to enter all the X-coordinate values of a nav_msgs/Path in x_set but I can't get it. 

I try it in different ways but I always end up with this mistake: 

```
terminate called after throwing an instance of' ecl:: StandardException'.
  what ():  
Location: /home/user/catkin_ws/src/src/ecl_core/ecl_geometry/src/lib/cubic_spline_blueprints. cpp: 89 
Flag: Out of range error - tried to access beyond the range of the object (usually container).
```
What is the right way to do it? Thank you

**Comment by stonier** [comment]
Your backtrace points to this [code](https://github.com/stonier/ecl_core/blob/devel/ecl_geometry/src/lib/cubic_spline_blueprints.cpp#L87-L89). Are you filling both `x_set` and `y_set` appropriately?

Do you have example code that is failing?

**Comment by hecperleo** [comment]
Well the answer is no, I'm not doing it the right way.
x_set is defined as ecl:: Array<double> x_set (wp_total); where wp_total is poseListX. size ().

posListX is defined as std:: vector<double>
poseX is defined as double. 
```
void UALPathCallback(const nav_msgs::Path& msg){
  msgDrawPath.header.frame_id = "map";
  msgDrawPath.poses = msg.poses;
    for(auto p : msg.poses){
      waypoint.pose.position.x = p.pose.position.x;
      waypoint.pose.position.y = p.pose.position.y;
      waypoint.pose.position.z = p.pose.position.z;
      poseList.push_back(waypoint);
      poseX = p.pose.position.x;
      poseListX.push_back(poseX);
      poseY = p.pose.position.y;
      poseListY.push_back(poseY);
      poseZ = p.pose.position.z;
      poseListZ.push_back(poseZ);
    }
```
Then
```
    double * wp_x = (double*)malloc(sizeof(double)*wp_total);
    double * wp_y = (double*)malloc(sizeof(double)*wp_total);
    double * wp_z = (double*)malloc(sizeof(double)*wp_total);

    for(int i=0; i<wp_total; i++){
      wp_x[i]=poseListX[i];
      wp_y[i]=poseListY[i];
      wp_y[i]=poseListZ[i];
    }
    double * x_ptr;
    double * y_ptr;
    double * z_ptr;
    x_ptr = wp_x;
    y_ptr = wp_y;
    z_ptr = wp_z;
    ecl::CubicSpline spline_x;
    ecl::CubicSpline spline_y;
    ecl::CubicSpline spline_z;
      for (int i = 0; i < wp_total; i++)
      {
        t_set[i] = (double)i;
        x_set[i] = x_ptr[i];
        y_set[i] = y_ptr[i];
        z_set[i] = z_ptr[i];
      }
      spline_x = ecl::CubicSpline::Natural(t_set,x_set);
      spline_y = ecl::CubicSpline::Natural(t_set,y_set);
      spline_z = ecl::CubicSpline::Natural(t_set,z_set);
```
Thank you for responding so quickly.

**Comment by stonier** [comment]
Can you check t_set, x_set, y_set and z_set sizes at the time of the exception?

**Comment by hecperleo** [comment]
All three are zero. 

If instead of subscribing to a topic with nav_msgs:: path I enter a path as follows: 

```
    wp_x[0] = 10; wp_x[1] = 12.5; wp_x[2] = 12.5; wp_x[3] = 12.5;
    wp_y[0] = 10; wp_y[1] = 10; wp_y[2] = 12.5; wp_y[3] = 15;
    wp_x[4] = 15; wp_x[5] = 17.5; wp_x[6] = 17.5; wp_x[7] = 17.5;
    wp_y[4] = 15; wp_y[5] = 15; wp_y[6] = 17.5; wp_y[7] = 20;
    wp_x[8] = 20; wp_x[9] = 22.5; wp_x[10] = 22.5; wp_x[11] = 22.5;
    wp_y[8] = 20; wp_y[9] = 20; wp_y[10] = 22.5; wp_y[11] = 25;
    int wp_total = 12;

```
It works perfectly without touching up anything else from the code. But as i said before, i need to suscribe to a topic. 

By the way, i checked this topic and its working fine, thats why I think that I am making mistakes with variable types.

**Comment by hecperleo** [comment]
I found the error. I was reading the topic well with nav_msgs/Path but I wasn't saving its variables well so the size of x_set. y_set and z_set gave me zero. 

Thank you so much for the help.

**Comment by hecperleo** [comment]
Hi, I don't know if I am using the issue system correctly but being the same person and the question related to the previous one, I reopen the issue. 

Is it possible to change the curvature of the spline? Theoretically it is supposed that modifying the values of t_set should change the curvature, I have tried but without success. 

I leave a picture of my 3D path today: 
![rviz_screenshot_2018_01_12-11_41_02](https://user-images.githubusercontent.com/25062331/34872153-2b88f90e-f790-11e7-8196-19e09127249c.png)

I think I should modify the values of t_set in this part, instead of going from 1 to 1, that goes from 0.5 to 0.5 but I can't get it to modify the curvature. The code part is as follows: 

```
ecl:: Array<double> t_set (wp_total);
ecl:: Array<double> x_set (wp_total);
ecl:: Array<double> y_set (wp_total);
ecl:: Array<double> z_set (wp_total);
for (int i = 0; i < wp_total; i++)
{
	x_set[i] = x_ptr[i];
	y_set[i] = y_ptr[i];
	z_set[i] = z_ptr[i];
	t_set[i] = (double)i;
}

```
Basically what I want to do, which I don't know if it is possible, is to give the function run times on each waypoint. In this way the curvature would have to be more or less aggressive to meet those times. 

Thank you very much.

**Comment by stonier** [comment]
If you're using cubic splines, all of the parameters are tuned to meet the needs of smoothing. You'd need more parameters in your spline to dedicate it to changing the curvature. Even then, the results of that tuning are surprising rather than controllable to some mathematical metrics.

This is where I typically use the tension splines, see the documentation [here](http://docs.ros.org/kinetic/api/ecl_geometry/html/splinesGeometry.html). They give you some control over the shape of the curve in a non-surprising way that can even be tuned to meet acceleration limits.
